File: src/event-graph/graph-helpers.ts
/**
 * Event Graph — Graph Helpers
 *
 * Pure functions for manipulating the requires/provides task dependency graph.
 * No I/O, no side effects.
 */

import type { GraphConfig, TaskConfig, GraphEngineStore, ExecutionState, RefreshStrategy } from './types.js';
import { TASK_STATUS } from './constants.js';

// ============================================================================
// Accessors — normalize requires/provides to always be arrays
// ============================================================================

export function getProvides(task: TaskConfig | undefined): string[] {
  if (!task) return [];
  if (Array.isArray(task.provides)) return task.provides;
  return [];
}

export function getRequires(task: TaskConfig | undefined): string[] {
  if (!task) return [];
  if (Array.isArray(task.requires)) return task.requires;
  return [];
}

export function getAllTasks(graph: GraphConfig): Record<string, TaskConfig> {
  return graph.tasks ?? {};
}

export function getTask(graph: GraphConfig, taskName: string): TaskConfig | undefined {
  return graph.tasks[taskName];
}

export function hasTask(graph: GraphConfig, taskName: string): boolean {
  return taskName in graph.tasks;
}

// ============================================================================
// Task State Predicates
// ============================================================================

export function isNonActiveTask(taskState: GraphEngineStore | undefined): boolean {
  if (!taskState) return false;
  return taskState.status === TASK_STATUS.FAILED || taskState.status === TASK_STATUS.INACTIVATED;
}

export function isTaskCompleted(taskState: GraphEngineStore | undefined): boolean {
  return taskState?.status === TASK_STATUS.COMPLETED;
}

export function isTaskRunning(taskState: GraphEngineStore | undefined): boolean {
  return taskState?.status === TASK_STATUS.RUNNING;
}

export function getRefreshStrategy(taskConfig: TaskConfig, graphSettings?: { refreshStrategy?: RefreshStrategy }): RefreshStrategy {
  return taskConfig.refreshStrategy ?? graphSettings?.refreshStrategy ?? 'data-changed';
}

export function isRerunnable(taskConfig: TaskConfig, graphSettings?: { refreshStrategy?: RefreshStrategy }): boolean {
  return getRefreshStrategy(taskConfig, graphSettings) !== 'once';
}

export function getMaxExecutions(taskConfig: TaskConfig): number | undefined {
  return taskConfig.maxExecutions;
}

// ============================================================================
// Available Outputs Computation
// ============================================================================

/**
 * Dynamically compute available outputs from all completed tasks.
 * Tasks with strategies other than 'once' may have completed and reset.
 * Pure function.
 */
export function computeAvailableOutputs(
  graph: GraphConfig,
  taskStates: Record<string, GraphEngineStore>
): string[] {
  const outputs: Set<string> = new Set();

  for (const [taskName, taskState] of Object.entries(taskStates)) {
    if (taskState.status === TASK_STATUS.COMPLETED) {
      const taskConfig = graph.tasks[taskName];
      if (taskConfig) {
        const provides = getProvides(taskConfig);
        provides.forEach(output => outputs.add(output));
      }
    }
  }

  return Array.from(outputs);
}

// ============================================================================
// Conflict Detection
// ============================================================================

/**
 * Group candidate tasks by the outputs they provide.
 * Used to detect conflicts (multiple tasks providing the same output).
 */
export function groupTasksByProvides(
  candidateTaskNames: string[],
  tasks: Record<string, TaskConfig>
): Record<string, string[]> {
  const outputGroups: Record<string, string[]> = {};

  candidateTaskNames.forEach(taskName => {
    const task = tasks[taskName];
    if (!task) return;
    const provides = getProvides(task);
    provides.forEach(output => {
      if (!outputGroups[output]) {
        outputGroups[output] = [];
      }
      outputGroups[output].push(taskName);
    });
  });

  return outputGroups;
}

/**
 * Check if a task's outputs conflict with other candidates.
 */
export function hasOutputConflict(
  taskName: string,
  taskProvides: string[],
  candidates: string[],
  tasks: Record<string, TaskConfig>
): boolean {
  for (const otherName of candidates) {
    if (otherName === taskName) continue;
    const otherProvides = getProvides(tasks[otherName]);
    const overlapping = taskProvides.some(output => otherProvides.includes(output));
    if (overlapping) return true;
  }
  return false;
}

// ============================================================================
// Immutable Graph Mutation
// ============================================================================

export function addKeyToProvides(task: TaskConfig, key: string): TaskConfig {
  const current = getProvides(task);
  if (current.includes(key)) return task;
  return { ...task, provides: [...current, key] };
}

export function removeKeyFromProvides(task: TaskConfig, key: string): TaskConfig {
  const current = getProvides(task);
  return { ...task, provides: current.filter(p => p !== key) };
}

export function addKeyToRequires(task: TaskConfig, key: string): TaskConfig {
  const current = getRequires(task);
  if (current.includes(key)) return task;
  return { ...task, requires: [...current, key] };
}

export function removeKeyFromRequires(task: TaskConfig, key: string): TaskConfig {
  const current = getRequires(task);
  return { ...task, requires: current.filter(r => r !== key) };
}

// ============================================================================
// Dynamic Task Management
// ============================================================================

/**
 * Add a new task to a graph config. Returns a new GraphConfig (immutable).
 */
export function addDynamicTask(
  graph: GraphConfig,
  taskName: string,
  taskConfig: TaskConfig
): GraphConfig {
  return {
    ...graph,
    tasks: {
      ...graph.tasks,
      [taskName]: taskConfig,
    },
  };
}

/**
 * Create default task state for a new task.
 */
export function createDefaultGraphEngineStore(): GraphEngineStore {
  return {
    status: 'not-started',
    executionCount: 0,
    retryCount: 0,
    lastEpoch: 0,
    messages: [],
    progress: null,
  };
}

/**
 * Create the initial execution state for a graph.
 */
export function createInitialExecutionState(
  graph: GraphConfig,
  executionId: string
): ExecutionState {
  const tasks: Record<string, GraphEngineStore> = {};
  for (const taskName of Object.keys(graph.tasks)) {
    tasks[taskName] = createDefaultGraphEngineStore();
  }

  return {
    status: 'running',
    tasks,
    availableOutputs: [],
    stuckDetection: { is_stuck: false, stuck_description: null, outputs_unresolvable: [], tasks_blocked: [] },
    lastUpdated: new Date().toISOString(),
    executionId,
    executionConfig: {
      executionMode: graph.settings.execution_mode ?? 'eligibility-mode',
      conflictStrategy: graph.settings.conflict_strategy ?? 'alphabetical',
      completionStrategy: graph.settings.completion,
    },
  };
}

File: pycli/pylib/event_graph/graph_helpers.py
"""
Event Graph — Graph Helpers

Pure functions for manipulating the requires/provides task dependency graph.
No I/O, no side effects.

Port of src/event-graph/graph-helpers.ts
"""
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

from .types import TASK_STATUS


# ============================================================================
# Accessors — normalize requires/provides to always be lists
# ============================================================================

def get_provides(task: dict | None) -> list[str]:
    if not task:
        return []
    provides = task.get("provides")
    return list(provides) if isinstance(provides, list) else []


def get_requires(task: dict | None) -> list[str]:
    if not task:
        return []
    requires = task.get("requires")
    return list(requires) if isinstance(requires, list) else []


def get_all_tasks(graph: dict) -> dict:
    return graph.get("tasks") or {}


def get_task(graph: dict, task_name: str) -> dict | None:
    return graph.get("tasks", {}).get(task_name)


def has_task(graph: dict, task_name: str) -> bool:
    return task_name in graph.get("tasks", {})


# ============================================================================
# Task State Predicates
# ============================================================================

def is_non_active_task(task_state: dict | None) -> bool:
    if not task_state:
        return False
    return task_state.get("status") in (TASK_STATUS["FAILED"], TASK_STATUS["INACTIVATED"])


def is_task_completed(task_state: dict | None) -> bool:
    if not task_state:
        return False
    return task_state.get("status") == TASK_STATUS["COMPLETED"]


def is_task_running(task_state: dict | None) -> bool:
    if not task_state:
        return False
    return task_state.get("status") == TASK_STATUS["RUNNING"]


def get_refresh_strategy(task_config: dict, graph_settings: dict | None = None) -> str:
    strategy = task_config.get("refreshStrategy")
    if strategy:
        return strategy
    if graph_settings:
        strategy = graph_settings.get("refreshStrategy")
        if strategy:
            return strategy
    return "data-changed"


def is_rerunnable(task_config: dict, graph_settings: dict | None = None) -> bool:
    return get_refresh_strategy(task_config, graph_settings) != "once"


def get_max_executions(task_config: dict) -> int | None:
    return task_config.get("maxExecutions")


# ============================================================================
# Available Outputs Computation
# ============================================================================

def compute_available_outputs(graph: dict, task_states: dict) -> list[str]:
    """
    Dynamically compute available outputs from all completed tasks.
    Pure function.
    """
    outputs: set[str] = set()

    for task_name, task_state in task_states.items():
        if task_state.get("status") == TASK_STATUS["COMPLETED"]:
            task_config = graph.get("tasks", {}).get(task_name)
            if task_config:
                for output in get_provides(task_config):
                    outputs.add(output)

    return list(outputs)


# ============================================================================
# Conflict Detection
# ============================================================================

def group_tasks_by_provides(
    candidate_task_names: list[str],
    tasks: dict,
) -> dict[str, list[str]]:
    """
    Group candidate tasks by the outputs they provide.
    Used to detect conflicts.
    """
    output_groups: dict[str, list[str]] = {}

    for task_name in candidate_task_names:
        task = tasks.get(task_name)
        if not task:
            continue
        for output in get_provides(task):
            if output not in output_groups:
                output_groups[output] = []
            output_groups[output].append(task_name)

    return output_groups


def has_output_conflict(
    task_name: str,
    task_provides: list[str],
    candidates: list[str],
    tasks: dict,
) -> bool:
    """Check if a task's outputs conflict with other candidates."""
    for other_name in candidates:
        if other_name == task_name:
            continue
        other_provides = get_provides(tasks.get(other_name))
        for output in task_provides:
            if output in other_provides:
                return True
    return False


# ============================================================================
# Default state factories
# ============================================================================

def create_default_graph_engine_store() -> dict:
    """Create default task state for a new task."""
    return {
        "status": "not-started",
        "executionCount": 0,
        "retryCount": 0,
        "lastEpoch": 0,
        "messages": [],
        "progress": None,
    }


def create_initial_execution_state(graph: dict, execution_id: str) -> dict:
    """Create the initial execution state for a graph."""
    tasks = {}
    for task_name in graph.get("tasks", {}).keys():
        tasks[task_name] = create_default_graph_engine_store()

    settings = graph.get("settings", {})

    return {
        "status": "running",
        "tasks": tasks,
        "availableOutputs": [],
        "stuckDetection": {
            "is_stuck": False,
            "stuck_description": None,
            "outputs_unresolvable": [],
            "tasks_blocked": [],
        },
        "lastUpdated": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z",
        "executionId": execution_id,
        "executionConfig": {
            "executionMode": settings.get("execution_mode", "eligibility-mode"),
            "conflictStrategy": settings.get("conflict_strategy", "alphabetical"),
            "completionStrategy": settings.get("completion", "manual"),
        },
    }
